File: gitpane/watcher.py
"""Filesystem watching that reports repository invalidations."""

import asyncio
from collections.abc import AsyncIterator, Iterable
from dataclasses import dataclass
from pathlib import Path

from watchfiles import awatch

from gitpane import git

DEBOUNCE_MS = 250


@dataclass(frozen=True)
class Invalidation:
    """What a batch of filesystem events made stale."""

    status: bool = False
    history: bool = False
    changed_paths: frozenset[Path] = frozenset()
    index_changed: bool = False

    def merge(self, other: "Invalidation") -> "Invalidation":
        """Return an invalidation covering both *self* and *other*."""
        return Invalidation(
            self.status or other.status,
            self.history or other.history,
            self.changed_paths | other.changed_paths,
            self.index_changed or other.index_changed,
        )


def classify(
    root: Path, git_dir: Path, common_dir: Path, paths: Iterable[Path]
) -> Invalidation:
    """Classify changed absolute *paths* for the worktree at *root*.

    ``changed_paths`` holds worktree paths relative to *root*. Only the index,
    HEAD, refs and packed refs are relevant below the Git directories.
    """
    result = Invalidation()
    for path in paths:
        if path.is_relative_to(git_dir):
            name = path.relative_to(git_dir)
            if name == Path("index"):
                result = result.merge(Invalidation(status=True, index_changed=True))
                continue
            if name == Path("HEAD"):
                result = result.merge(Invalidation(status=True, history=True))
                continue
        if path.is_relative_to(common_dir):
            name = path.relative_to(common_dir)
            if name == Path("packed-refs") or name.parts[:1] == ("refs",):
                result = result.merge(Invalidation(status=True, history=True))
            continue
        if path == root / ".git":
            continue
        if path.is_relative_to(root):
            result = result.merge(
                Invalidation(
                    status=True, changed_paths=frozenset({path.relative_to(root)})
                )
            )
    return result


async def watch(root: Path) -> AsyncIterator[Invalidation]:
    """Yield debounced invalidations for the repository at *root*.

    Setup failures such as ``OSError`` propagate to the caller.
    """
    git_dir, common_dir = await asyncio.to_thread(git.git_dirs, root)
    roots = {root}
    roots.update(d for d in (git_dir, common_dir) if not d.is_relative_to(root))
    # The default filter drops any path containing ".git" (and cache dirs),
    # which would hide the metadata events; classify() is the only filter.
    async for changes in awatch(*roots, debounce=DEBOUNCE_MS, watch_filter=None):
        invalidation = classify(
            root, git_dir, common_dir, (Path(path) for _, path in changes)
        )
        if invalidation != Invalidation():
            yield invalidation

File: src/watcher.rs
//! Filesystem watching that reports repository invalidations.

use std::collections::BTreeSet;
use std::path::{Path, PathBuf};
use std::sync::Arc;
use std::sync::mpsc::{self, Sender};
use std::thread;
use std::time::Duration;

use notify_debouncer_full::new_debouncer;
use notify_debouncer_full::notify::RecursiveMode;

use crate::app::Event;
use crate::git::GitApi;

const DEBOUNCE: Duration = Duration::from_millis(250);

/// What a batch of filesystem events made stale.
#[derive(Debug, Clone, Default, PartialEq, Eq)]
pub struct Invalidation {
    pub status: bool,
    pub history: bool,
    /// Worktree paths relative to the root, `/`-separated.
    pub changed_paths: BTreeSet<String>,
    pub index_changed: bool,
}

impl Invalidation {
    /// Extend `self` to cover `other` as well.
    pub fn merge(&mut self, other: Invalidation) {
        self.status |= other.status;
        self.history |= other.history;
        self.changed_paths.extend(other.changed_paths);
        self.index_changed |= other.index_changed;
    }
}

/// Watcher lifecycle and change messages.
#[derive(Debug)]
pub enum Watch {
    /// Watching succeeded.
    Started,
    Changed(Invalidation),
    /// Setup failed (before `Started`), or watching ended or failed after it.
    Stopped,
}

/// Classify changed absolute `paths` for the worktree at `root`.
pub fn classify<'a>(
    root: &Path,
    git_dir: &Path,
    common_dir: &Path,
    paths: impl IntoIterator<Item = &'a Path>,
) -> Invalidation {
    let mut result = Invalidation::default();
    for path in paths {
        if let Ok(name) = path.strip_prefix(git_dir) {
            if name == Path::new("index") {
                result.status = true;
                result.index_changed = true;
                continue;
            }
            if name == Path::new("HEAD") {
                result.status = true;
                result.history = true;
                continue;
            }
        }
        if let Ok(name) = path.strip_prefix(common_dir) {
            if name == Path::new("packed-refs") || name.starts_with("refs") {
                result.status = true;
                result.history = true;
            }
            continue;
        }
        if path == root.join(".git") {
            continue;
        }
        if let Ok(relative) = path.strip_prefix(root) {
            result.status = true;
            result
                .changed_paths
                .insert(relative.to_string_lossy().into_owned());
        }
    }
    result
}

/// Watch the repository at `root` on a thread, posting `Event::Watch` messages.
pub fn spawn(git: Arc<dyn GitApi>, root: PathBuf, tx: Sender<Event>) {
    thread::spawn(move || {
        let _ = watch(git.as_ref(), &root, &tx);
        let _ = tx.send(Event::Watch(Watch::Stopped));
    });
}

/// Run the watcher until it fails, ends, or the app is gone.
fn watch(git: &dyn GitApi, root: &Path, tx: &Sender<Event>) -> Result<(), ()> {
    let (git_dir, common_dir) = git.git_dirs(root).map_err(|_| ())?;
    let (events, rx) = mpsc::channel();
    let mut debouncer = new_debouncer(DEBOUNCE, None, events).map_err(|_| ())?;
    debouncer
        .watch(root, RecursiveMode::Recursive)
        .map_err(|_| ())?;
    for dir in [&git_dir, &common_dir] {
        if !dir.starts_with(root) {
            debouncer
                .watch(dir, RecursiveMode::Recursive)
                .map_err(|_| ())?;
        }
    }
    tx.send(Event::Watch(Watch::Started)).map_err(|_| ())?;
    for batch in rx {
        let batch = batch.map_err(|_| ())?;
        let paths = batch.iter().flat_map(|event| event.paths.iter());
        let invalidation = classify(root, &git_dir, &common_dir, paths.map(PathBuf::as_path));
        if invalidation != Invalidation::default() {
            tx.send(Event::Watch(Watch::Changed(invalidation)))
                .map_err(|_| ())?;
        }
    }
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    const ROOT: &str = "/work/repo";
    const LINKED_ROOT: &str = "/work/linked";
    const LINKED_GIT: &str = "/work/repo/.git/worktrees/linked";
    const COMMON: &str = "/work/repo/.git";

    fn normal(paths: &[&str]) -> Invalidation {
        let paths: Vec<&Path> = paths.iter().map(Path::new).collect();
        classify(Path::new(ROOT), Path::new(COMMON), Path::new(COMMON), paths)
    }

    fn linked(paths: &[&str]) -> Invalidation {
        let paths: Vec<&Path> = paths.iter().map(Path::new).collect();
        classify(
            Path::new(LINKED_ROOT),
            Path::new(LINKED_GIT),
            Path::new(COMMON),
            paths,
        )
    }

    fn edited(paths: &[&str]) -> Invalidation {
        Invalidation {
            status: true,
            changed_paths: paths.iter().map(|p| p.to_string()).collect(),
            ..Invalidation::default()
        }
    }

    fn index() -> Invalidation {
        Invalidation {
            status: true,
            index_changed: true,
            ..Invalidation::default()
        }
    }

    fn history() -> Invalidation {
        Invalidation {
            status: true,
            history: true,
            ..Invalidation::default()
        }
    }

    #[test]
    fn worktree_edit_invalidates_status_with_relative_path() {
        assert_eq!(normal(&["/work/repo/src/a.py"]), edited(&["src/a.py"]));
    }

    #[test]
    fn index_change_invalidates_status_and_index() {
        assert_eq!(normal(&["/work/repo/.git/index"]), index());
    }

    #[test]
    fn head_refs_and_packed_refs_invalidate_status_and_history() {
        for name in ["HEAD", "refs/heads/main", "packed-refs"] {
            let path = format!("{COMMON}/{name}");
            assert_eq!(normal(&[&path]), history(), "{name}");
        }
    }

    #[test]
    fn other_git_metadata_is_ignored() {
        let paths = [
            "/work/repo/.git/logs/HEAD",
            "/work/repo/.git/objects/ab/cdef",
            "/work/repo/.git/index.lock",
            "/work/repo/.git/config",
        ];
        assert_eq!(normal(&paths), Invalidation::default());
    }

    #[test]
    fn linked_worktree_splits_metadata_between_directories() {
        assert_eq!(linked(&[&format!("{LINKED_GIT}/index")]), index());
        assert_eq!(linked(&[&format!("{LINKED_GIT}/HEAD")]), history());
        assert_eq!(linked(&[&format!("{COMMON}/refs/heads/main")]), history());
        assert_eq!(linked(&[&format!("{COMMON}/packed-refs")]), history());
    }

    #[test]
    fn linked_worktree_ignores_other_metadata_and_git_file() {
        let paths = [
            "/work/linked/.git".to_string(),
            format!("{LINKED_GIT}/logs/HEAD"),
            format!("{COMMON}/index"),
            format!("{COMMON}/HEAD"),
            format!("{COMMON}/objects/ab"),
        ];
        let paths: Vec<&str> = paths.iter().map(String::as_str).collect();
        assert_eq!(linked(&paths), Invalidation::default());
    }

    #[test]
    fn paths_outside_repository_are_ignored() {
        assert_eq!(normal(&["/elsewhere/file"]), Invalidation::default());
    }

    #[test]
    fn batch_and_merge_combine_invalidations() {
        let mut merged = normal(&[
            "/work/repo/a.txt",
            "/work/repo/.git/index",
            "/work/repo/.git/HEAD",
        ]);
        merged.merge(Invalidation {
            changed_paths: ["b.txt".to_string()].into(),
            ..Invalidation::default()
        });
        assert_eq!(
            merged,
            Invalidation {
                status: true,
                history: true,
                changed_paths: ["a.txt".to_string(), "b.txt".to_string()].into(),
                index_changed: true,
            }
        );
    }
}
